File: 15_sum_In_Range/sumInRange.py
# Sum In Range - Interview Question - Asked by Palantir Technologies
# In Python
# In O(n) time

def sumInRange(nums, queries):
    totalSum = 0
    prefixSums = []
    modulo = 1000000000 + 7
    
    # calculates the prefix-sums
    for i in range(len(nums)):
        if (i == 0):
            prefixSums.append(nums[i] % modulo)
        else:
            prefixSums.append((prefixSums[i-1] + nums[i]) % modulo)

    # Now adds up each separate range
    for query in queries:
        fromIndex = query[0]
        toIndex = query[1]
        totalSum += prefixSums[toIndex]

        if (fromIndex != 0): totalSum -= prefixSums[fromIndex-1]
        totalSum %= modulo
    # if totalSum is negative we just add the modulo
    if (totalSum < 0): totalSum += modulo
    return totalSum

# Example
nums =  [34, 19, 21, 5, 1, 10, 26, 46, 33, 10]
queries = [[3,7], 
 [3,4], 
 [3,7], 
 [4,5], 
 [0,5]]

print(sumInRange(nums, queries)) # 283

File: 15_sum_In_Range/sumInRange.go
package main

import "fmt"

// Sum In Range - Interview Question - Asked by Palantir Technologies
// In Golang
// In O(n) time
func sumInRange(nums []int, queries [][]int) int {
	var totalSum int
	var prefixSums = map[int]int{}
	modulo := 1000000000 + 7

	// calculates the prefix-sums
	for index, num := range nums {
		if index == 0 {
			prefixSums[index] = num % modulo
		} else {
			prefixSums[index] = (prefixSums[index-1] + num) % modulo
		}
	}

	// Now adds up each separate range
	for _, query := range queries {
		fromIndex := query[0]
		toIndex := query[1]
		totalSum += prefixSums[toIndex]

		if fromIndex != 0 {
			totalSum -= prefixSums[fromIndex-1]
		}
		totalSum %= modulo
	}
	// if totalSum is negative we just add the modulo
	if totalSum < 0 {
		totalSum += modulo
	}
	return totalSum
}

func main() {
	// Example
	nums := []int{34, 19, 21, 5, 1, 10, 26, 46, 33, 10}
	queries := [][]int{{3, 7},
		{3, 4},
		{3, 7},
		{4, 5},
		{0, 5}}

	fmt.Println(sumInRange(nums, queries)) // 283
}
